### PR TITLE
feat(microservice): add TLS over TCP support

### PR DESCRIPTION
**Pull request by Flusinerd**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
Adds an additional property ``useTls?: boolean`` to the options for client and server. If useTls is set, it intersects the Nodes tls server / client options with the nest server/client options.

If useTls is set to true, it uses node's ``tls`` module instead of ``net``  to allow TCP over TLS support

```
[ ] Bugfix
[ X ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #6745 

## What is the new behavior?

If ``useTls === true`` then it will use ``tls`` instead of ``net`` to create the server / client and intersect the appropriate options with the nest options  

## Does this PR introduce a breaking change?
```
[ ] Yes
[ X ] No
```

## Other information

### CODE REVIEW

**Comment by Flusinerd** [comment]
For docs, should I update the docs in ``nestjs/docs.nestjs.com``?

I think we could either add another section in:
https://docs.nestjs.com/microservices/basics#client

Or make an own chapter for ``TCP Client`` or ``TCP TLS Client``

**Comment by kamilmysliwiec** [comment]
> For docs, should I update the docs in nestjs/docs.nestjs.com?

Correct

> I think we could either add another section in:

Let's add a new section in the TCP chapter

**Comment by Flusinerd** [comment]
Docs have been updated as well, the PR is linked above

**Comment by Flusinerd** [comment]
I can add a sample later that I used for testing, should I include a self-signed certificate in the sample?
I'm waiting for #7486 to merged, since this includes support for multi-application samples.

File: packages/microservices/client/client-proxy-factory.ts [hunk 1]
```diff
@@ -56,13 +57,23 @@ export class ClientProxyFactory {
...
+          | TcpClientOptions['options']
+          | TcpTlsClientOptions['options']
+          | undefined;
+        if (uncheckedOptions && uncheckedOptions.useTls === true) {
+          return new ClientTCP(options as TcpTlsClientOptions['options']);
+        } else {
+          return new ClientTCP(
+            options as TcpClientOptions['options'] | undefined,
+          );
+        }
```
Comment (kamilmysliwiec) [inline packages/microservices/client/client-proxy-factory.ts:70]: As this condition seems to be used only for type-checking purposes, can we just force cast `options` to the appropriate type and remove this `if`?

File: packages/microservices/client/client-tcp.ts [hunk 2]
```diff
@@ -1,32 +1,55 @@
...
+  constructor(options: TcpTlsClientOptions['options']);
+  constructor(
+    private readonly options?:
+      | TcpClientOptions['options']
+      | TcpTlsClientOptions['options'],
+  ) {
     super();
+    if (options === undefined) {
+      this.options = {};
+    }
```
Comment (kamilmysliwiec) [inline packages/microservices/client/client-tcp.ts:49]: What's the reason for this change?

File: packages/microservices/test/client/client-tcp.spec.ts [hunk 3]
```diff
@@ -197,4 +199,15 @@ describe('ClientTCP', () => {
...
+  // describe('tls', () => {
+  //   it('should upgrade to TLS', () => {
+  //     const jsonSocket = new ClientTCP({ useTls: true }).createSocket();
+  //     expect(jsonSocket.socket).instanceOf(TLSSocket);
+  //   });
+  //   it('should not upgrade to TLS, if not requested', () => {
+  //     const jsonSocket = new ClientTCP({ useTls: false }).createSocket();
+  //     expect(jsonSocket.socket).instanceOf(NetSocket);
+  //   });
+  // });
```
Comment (kamilmysliwiec) [inline packages/microservices/test/client/client-tcp.spec.ts:212]: Is this intentionally commented out?

**Comment by kamilmysliwiec** [comment]
I left a few comments. Can we also add an integration test?

**Comment by Flusinerd** [comment]
Hi, ty for the the feedback I'll try work on this this weekend.

File: packages/microservices/client/client-proxy-factory.ts [hunk 4]
```diff
@@ -56,13 +57,23 @@ export class ClientProxyFactory {
       case Transport.KAFKA:
         return new ClientKafka(options as KafkaOptions['options']);
       default:
-        return new ClientTCP(options as TcpClientOptions['options']);
+        const uncheckedOptions = options as
```
Comment (AnisJS09) [inline packages/microservices/client/client-proxy-factory.ts:60]:
```suggestion
       const type UncheckedOptionsType = 
          | TcpClientOptions['options']
          | TcpTlsClientOptions['options']
          | undefined;
        const uncheckedOptions = options as UncheckedOptionsType
```

File: packages/microservices/client/client-proxy.ts [hunk 5]
```diff
@@ -140,26 +141,32 @@ export abstract class ClientProxy {
...
-        (options as
-          | RedisOptions['options']
-          | NatsOptions['options']
-          | MqttOptions['options']
-          | TcpClientOptions['options']
-          | RmqOptions['options']
-          | KafkaOptions['options']).serializer) ||
+        (
+          options as
+            | RedisOptions['options']
```
Comment (AnisJS09) [inline packages/microservices/client/client-proxy.ts:146]: what do you think about extracting these types into a Union type declared, something like : 
```suggestion
            const type Options = NatsOptions['options']
            | MqttOptions['options']
            | TcpClientOptions['options']
            | TcpTlsClientOptions['options']
            | RmqOptions['options']
            | KafkaOptions['options'];
```

**Comment by Flusinerd** [comment]
Currently I dont have that much time to work on this @AnisJS09 Do you want to continue?

**Comment by Flusinerd** [comment]
Okay, I managed to get some time to continue working on this.
I worked on the review from @kamilmysliwiec and think I got it sorted. 

Looking into integration testing right now. It seems like there are no e2e tests for TCP  implemented atm.
So what do you want to be covered for TCP clients?
Just a simple addition using sync / async / rxjs of and rxjs from + exception and pub / sub.
Similar to how its done in the nats example?

**Comment by kamilmysliwiec** [comment]
TCP tests https://github.com/nestjs/nest/blob/master/integration/microservices/e2e/sum-rpc.spec.ts

**Comment by Flusinerd** [comment]
Ahh alright. I completely missed that. Will work on this later then and finish this feature request.

**Comment by Flusinerd** [comment]
Okay, I added the e2e tests.
I included a self signed certificate.

File: integration/microservices/e2e/sum-rpc-tls.spec.ts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+describe('RPC TLS transport', () => {
+  let server;
+  let app: INestApplication;
+  let key: string;
+  let cert: string;
+
+  before(() => {
+    // Generate a self-signed key pair
+    console.log(__dirname);
```
Comment (kamilmysliwiec) [inline integration/microservices/e2e/sum-rpc-tls.spec.ts:19]:
```suggestion
```

File: packages/microservices/client/client-tcp.ts [hunk 7]
```diff
@@ -39,17 +66,35 @@ export class ClientTCP extends ClientProxy {
...
-        );
-      }),
-      share(),
-    );
-
-    this.socket.connect(this.port, this.host);
+    let source$: Observable<any>;
+
+    if (this.useTls) {
+      this.netSocket.connect(this.port, this.host);
```
Comment (kamilmysliwiec) [inline packages/microservices/client/client-tcp.ts:72]:
```suggestion
```
this shouldn't be needed (see non-TLS path)

File: packages/microservices/client/client-tcp.ts [hunk 8]
```diff
@@ -39,17 +66,35 @@ export class ClientTCP extends ClientProxy {
...
+        tap(() => {
+          this.isConnected = true;
+          this.socket.on(MESSAGE_EVENT, (buffer: WritePacket & PacketId) =>
+            this.handleResponse(buffer),
+          );
+        }),
+        share(),
+      );
+      this.socket.connect(this.port, this.host);
+    }
```
Comment (kamilmysliwiec) [inline packages/microservices/client/client-tcp.ts:97]: If the only difference here is whether to use `netSocket` or `socket` (depending on whether TLS is on) we should probably refactor this logic to avoid repetitiveness.

**Comment by nomaxg** [comment]
@Flusinerd @kamilmysliwiec is this actively being worked on? I would be happy to pick it up

**Comment by Flusinerd** [comment]
Hi. I don't think so. Feel free to continue

**Comment by kamilmysliwiec** [comment]
Moving onto this PR https://github.com/nestjs/nest/pull/10628
